api/v1alpha1/validator/nicclusterpolicy_webhook.go

```diff
@@ -474,6 +484,168 @@ func validateRepository(repo string, allErrs field.ErrorList, fp *field.Path, ch
+type stateRenderData struct {
+	provider    containerResourcesProvider
+	newState    newStateFunc
+	manifestDir string
+}
+
+func (w *nicClusterPolicyValidator) validateContainerResources(
+	policy *v1alpha1.NicClusterPolicy, allErrs field.ErrorList) field.ErrorList {
+	fp := field.NewPath("spec")
+
+	manifestBaseDir := envConfig.ManifestBaseDir
+
+	states := map[string]stateRenderData{}
+
+	if policy.Spec.OFEDDriver != nil {
+		states["ofedDriver"] = stateRenderData{
+			policy.Spec.OFEDDriver, state.NewStateOFED,
+			filepath.Join(manifestBaseDir, "state-ofed-driver"),
+		}
+	}
+	if policy.Spec.RdmaSharedDevicePlugin != nil {
+		states["rdmaSharedDevicePlugin"] = stateRenderData{
+			policy.Spec.RdmaSharedDevicePlugin, state.NewStateSharedDp,
+			filepath.Join(manifestBaseDir, "state-rdma-device-plugin"),
+		}
+	}
+	if policy.Spec.SriovDevicePlugin != nil {
+		states["sriovDevicePlugin"] = stateRenderData{
+			policy.Spec.SriovDevicePlugin, state.NewStateSriovDp,
+			filepath.Join(manifestBaseDir, "state-sriov-device-plugin"),
+		}
+	}
+	if policy.Spec.IBKubernetes != nil {
+		states["ibKubernetes"] = stateRenderData{
+			policy.Spec.IBKubernetes, state.NewStateIBKubernetes,
+			filepath.Join(manifestBaseDir, "state-ib-kubernetes"),
+		}
+	}
+	if policy.Spec.NvIpam != nil {
+		states["nvIpam"] = stateRenderData{
+			policy.Spec.NvIpam, state.NewStateNVIPAMCNI,
+			filepath.Join(manifestBaseDir, "state-nv-ipam-cni"),
+		}
+	}
+	if policy.Spec.NicFeatureDiscovery != nil {
+		states["nicFeatureDiscovery"] = stateRenderData{
+			policy.Spec.NicFeatureDiscovery, state.NewStateNICFeatureDiscovery,
+			filepath.Join(manifestBaseDir, "state-nic-feature-discovery"),
+		}
+	}
+	for stateName, renderData := range states {
+		localData := renderData
+		allErrs = validateContainerResourcesIfNotNil(&localData, policy, allErrs, fp, stateName)
+	}
+
+	if policy.Spec.SecondaryNetwork != nil {
+		snfp := fp.Child("secondaryNetwork")
+
+		states := map[string]stateRenderData{}
+		if policy.Spec.SecondaryNetwork.CniPlugins != nil {
+			states["cniPlugins"] = stateRenderData{
+				policy.Spec.SecondaryNetwork.CniPlugins, state.NewStateCNIPlugins,
+				filepath.Join(manifestBaseDir, "state-container-networking-plugins"),
+			}
+		}
+		if policy.Spec.SecondaryNetwork.IPoIB != nil {
+			states["ipoib"] = stateRenderData{
+				policy.Spec.SecondaryNetwork.IPoIB, state.NewStateIPoIBCNI,
+				filepath.Join(manifestBaseDir, "state-ipoib-cni"),
+			}
+		}
+		if policy.Spec.SecondaryNetwork.Multus != nil {
+			states["multus"] = stateRenderData{
+				policy.Spec.SecondaryNetwork.Multus, state.NewStateMultusCNI,
+				filepath.Join(manifestBaseDir, "state-multus-cni"),
+			}
+		}
+		if policy.Spec.SecondaryNetwork.IpamPlugin != nil {
+			states["ipamPlugin"] = stateRenderData{
+				policy.Spec.SecondaryNetwork.IpamPlugin, state.NewStateWhereaboutsCNI,
+				filepath.Join(manifestBaseDir, "state-whereabouts-cni"),
+			}
+		}
+		for stateName, renderData := range states {
```

Comment:
is there a reason to do this in 2 steps ?

that is , why not populate states map with all states then loop over once.

Reply:
Because we want to provide the proper field path in the error message. Some states are configured under SecondaryNetwork section

Reply:
ack. so its either moving field path into the render data struct or keeping as is.

thx.